crates/project/src/search_history.rs

```diff
@@ -0,0 +1,258 @@
+/// Determines the behavior to use when inserting a new query into the search history.
+#[derive(Default, Debug, Clone, PartialEq)]
+pub enum QueryInsertionBehavior {
+    #[default]
+    /// Always insert the query to the search history.
+    AlwaysInsert,
+    /// Replace the previous query in the search history, if the new query contains the previous query.
+    ReplacePreviousIfContains,
+}
+
+/// A handle that maintains the current selection in the search history.
+/// This can be passed to the search history to update the selection accordingly,
+/// e.g. when using the up and down arrow keys to navigate the search history.
+#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
+pub struct SearchHistorySelectionHandle {
+    selection: Option<usize>,
+}
+
+impl SearchHistorySelectionHandle {
+    /// Resets the selection to `None`.
+    pub fn reset(&mut self) {
+        self.selection = None;
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SearchHistory {
+    history: Vec<String>,
+    max_history_len: Option<usize>,
+    insertion_behavior: QueryInsertionBehavior,
+}
+
+impl SearchHistory {
+    pub fn new(max_history_len: Option<usize>, insertion_behavior: QueryInsertionBehavior) -> Self {
+        SearchHistory {
+            max_history_len,
+            insertion_behavior,
+            history: Vec::new(),
+        }
+    }
+
+    pub fn add(&mut self, handle: &mut SearchHistorySelectionHandle, search_string: String) {
+        if let Some(selected_ix) = handle.selection {
+            if search_string == self.history[selected_ix] {
```

Comment:
Let's use `get` instead of `[]` here to avoid potential panics if some handle sends us  `Some(max_history_len + 1)` index for some reason?